### ISSUE TITLE
Automating update when building from source

### ISSUE THREAD

**Issue by tlhackque** [curl/curl#7756]
Building from source (safely and robustly) involves more that what's indicated on https://curl.se/docs/install.html.

After years of building cURL from a patchwork of manual steps and scripts, I decided to automate the entire process, which involves:

- Downloading a source release
- Verifying its signature with the proper release key
- configure, build, install, and verify
- cleanup
- handling various errors along the way

### I did this
Wrote an easily customizable `bash` script` that reduces the process to (after tailoring for your environment):
````sh
./curl-BUILD 7.79.1
````
If something goes wrong, you can restart at any time - e.g. if you need to tweak configure options or apply a patch:
````sh
cd curl-7.79.1
(fix...)
../curl-BUILD
````
(No, you don't need to pass the version string when inside a build directory...less typing is always better)

There are a couple of command options  for rare cases:
- stop after downloading and verifying the source archive
- keep files after a successful build (if you've enabled cleanup, which is off by default)

And if you use the same configure options everywhere, you can export a couple of variables so you don't have to edit the script:
- `PREFIX` - where to install
- `CURL` - Alternate download tool (with any options)

But in the usual case, it's just 20 keystrokes from receiving a release annoucement to installation.

### I expected the following
A very small/simple script
### I ended up with
Something a bit larger...

[curl-BUILD.tar.gz](https://github.com/curl/curl/files/7211782/curl-BUILD.tar.gz)

This should work any system that has `bash`; tested on Linux.
It does assume that you already have `cURL` installed - bootstrapping from a package distribution of `cURL` or specifying an alternate download tool is fairly straightforward.

I thought that since others might find it useful, you might want to include it in your "scripts" directory.  Any standard license (that includes the "use at your own risk" disclaimer) is fine.

Installation amounts to putting it in the directory where you download and expand your kits, adjusting `configure` options (and perhaps 3 script options).  Details in the script.  Hopefully, it will prove to be reasonably stable.

Enjoy.

**Comment by tlhackque** [comment]
Being a user is different from being a developer.  Our perspecitve is somewhat different.  We build cURL infrequently.  Everything isn't setup on disk - or in short-term memory.  Doing all the prework (that developers generally don't have to) **is** an issue; we make mistakes, w're slow, and it begged for automation.

I thought I should try (again) to give something back to the community.

I didn't create a pull request because (a) I don't currently build cURL from git; (b) wherever I put the script would be wrong (scripts? examples? contrib?); and (c) there's no point in my doing still more speculative work if the script is unwelcome - which it sounds like it is.

There's not much point in hosting it somewhere else - it wouldn't be found.  (And it's really too small to justify its own repo.)

You've already done a great job of automating the rest of the build process, so no surprise that this script simply calls it.  I needed to work on the stuff that comes first - which is more painful than it seems to a developer who has everything available locally.

Most people get the e-mail, go to the website, find the download, copy the iink, switch to a build or targe system, and do the dowload.  Then look around for the configure options that worked last time.  Did I keep the last build directory?  Which file?  Or an old-fashioned notebook?  Try configure.  Maybe iterate once or twice.  Then compile.  And it all gives rise to typos, redos, and cursing.  In my experience, people tend to skip verifying the package, which is increasingly important in our hostile world.  It's all work.  It's serial.  And I'd much rather set something up once, and after that type 20 characters and walk away.

Your other observations are helpful and fixable.  Thanks for taking the time to make them.

configure - as the script indicates, you're expected to make it fit your environment.  What's there isn't a recommendation.  It's just what I happen to use on the system where I first wrote the script.  The options could be removed in a release to prevent the impression that it is a recommendation.

Of course people can (and should) decide if they want to make in parallel - I have no clue how many processors they have.  And since I build it once/system/release - mostly on low-end systems, and configure takes most of the time, I generally don't bother.  (For things I develop, with 10s of builds a day, it's worth optimizing.)

Agree that a  sudo in front of make install would be useful to some (and harmless to all)

The current documentation is adequate for me.  I certainly could upgrade it if there's interest.  It could even go on the install page...

You're correct that invoking cURL at completion should use the PREFIX.  An oversight.  I'll fix that.

Bottom line: If the project (and in particular the users) would like the script to be made available, and if you let me know where, if not `scripts/` you'd like it (many projects have a `contrib/` directory...), I can produce a version & pull request that addresses your comments.

If there is not interest I won't bother.  At this point "it works for me", and I tried to share.

**Comment by bagder** [comment]
Just adding the script to the git repository won't make people magically learn about its existence and start to use it either. The storage location is not the key. The information, documentation and maintenance are rather what can make that happen - if there's a desire for it out there. I rather see that there's such a desire and use case shown *first*, then we can discuss actually hosting and maintaining it somewhere in the curl github organization.

**Comment by bagder** [comment]
Thanks for your interest and willingness to help us improve curl. But unless we see more user interest in us providing a script like this, I will insist that it is better hosted and provided out of the git main repository.
